File: backend-dotnet/SahayaSetu/Controllers/DonorController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using SahayaSetu.DTO;
using SahayaSetu.DTO.Volunteer; // Added back
using SahayaSetu.Services;

namespace SahayaSetu.Controllers
{
    [Route("api/[controller]")]
    [ApiController]
    public class DonorController : ControllerBase
    {
        private readonly IDonorService _donorService;

        public DonorController(IDonorService donorService)
        {
            _donorService = donorService;
        }

        // View all open requests
        [HttpGet("requests")]
        public async Task<IActionResult> GetAvailableRequests()
        {
            var result = await _donorService.GetAllOpenRequestsAsync();
            return Ok(result);
        }
        
        [HttpGet("requests-for-donor")]
        public async Task<IActionResult> GetAllOpenRequestsForDonor([FromQuery] int donorId)
        {
             var result = await _donorService.GetAllOpenRequestsForDonorAsync(donorId);
             return Ok(result);
        }

        [HttpGet("fundraisers")]
        public async Task<IActionResult> GetOpenFundraisers()
        {
            var result = await _donorService.GetOpenFundraisersAsync();
            return Ok(result);
        }

        // Donate money (Fundraiser)
        [HttpPost("requests/{requestId}/donate")]
        public async Task<IActionResult> DonateMoney(int requestId, [FromQuery] int donorId, [FromBody] DonationDto dto)
        {
            await _donorService.DonateToFundraiserAsync(donorId, requestId, dto);
            return Ok("Donation processed successfully");
        }

        [HttpPost("requests/{requestId}/volunteer")]
        public async Task<IActionResult> ApplyForVolunteer(int requestId, [FromQuery] int donorId)
        {
            await _donorService.ApplyForVolunteerAsync(donorId, requestId);
            return Ok("Volunteer application submitted");
        }

        [HttpPost("requests/{requestId}/fulfill-resource")]
        public async Task<IActionResult> FulfillResource(int requestId, [FromQuery] int donorId, [FromBody] ResourceFulfillDto dto)
        {
             await _donorService.FulfillResourceRequestAsync(donorId, requestId, dto.Quantity);
             return Ok("Resource contribution recorded");
        }
        
        [HttpGet("contributions")]
        public async Task<IActionResult> GetContributions([FromQuery] int donorId)
        {
             var result = await _donorService.GetDonorContributionsAsync(donorId);
             return Ok(result);
        }

    }
}

File: backend-java/sahayaSetu/src/main/java/com/sahayaSetu/controllers/DonorController.java
package com.sahayaSetu.controllers;

import com.sahayaSetu.dtos.DonationDto;
import com.sahayaSetu.dtos.ResourceFulfillmentDto;
import com.sahayaSetu.services.IDonorService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/api/donor")
@RequiredArgsConstructor
public class DonorController {

    private final IDonorService donorService;

    @GetMapping("/requests")
    public ResponseEntity<?> getAllOpenRequests() {
        return ResponseEntity.ok(donorService.getAllOpenRequests());
    }

    @GetMapping("/requests-for-donor")
    public ResponseEntity<?> getAllOpenRequestsForDonor(@RequestParam Long donorId) {
        return ResponseEntity.ok(donorService.getAllOpenRequestsForDonor(donorId));
    }

    @GetMapping("/fundraisers")
    public ResponseEntity<?> getOpenFundraisers() {
        return ResponseEntity.ok(donorService.getOpenFundraisers());
    }

    @PostMapping("/requests/{requestId}/donate")
    public ResponseEntity<?> donate(@PathVariable Long requestId,
            @RequestParam Long donorId,
            @Valid @RequestBody DonationDto dto) {
        donorService.donateToFundraiser(donorId, requestId, dto);
        return ResponseEntity.ok("Donation processed successfully");
    }

    @PostMapping("/requests/{requestId}/volunteer")
    public ResponseEntity<?> volunteer(@PathVariable Long requestId,
            @RequestParam Long donorId) {
        donorService.applyForVolunteer(donorId, requestId);
        return ResponseEntity.ok("Volunteer application submitted");
    }

    @PostMapping("/requests/{requestId}/fulfill-resource")
    public ResponseEntity<?> fulfillResource(@PathVariable Long requestId,
            @RequestParam Long donorId,
            @Valid @RequestBody ResourceFulfillmentDto dto) {
        donorService.fulfillResourceRequest(donorId, requestId, dto.getQuantity());
        return ResponseEntity.ok("Resource contribution recorded");
    }

    @GetMapping("/contributions")
    public ResponseEntity<?> getContributions(@RequestParam Long donorId) {
        try {
            if (donorId == null || donorId <= 0) {
                return ResponseEntity.badRequest().body("Invalid donor ID");
            }
            return ResponseEntity.ok(donorService.getDonorContributions(donorId));
        } catch (RuntimeException e) {
            return ResponseEntity.badRequest().body(e.getMessage());
        } catch (Exception e) {
            return ResponseEntity.status(500).body("Internal server error: " + e.getMessage());
        }
    }
}
